sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SetFns.java

```diff
@@ -0,0 +1,261 @@
+   * PCollection<String> right = p.apply(Create.of("1", "3", "4", "6"));
+   *
+   * PCollection<String> results =
+   *     left.apply(SetFns.union(right));
+   * }</pre>
+   */
+  public static <T> SetImpl<T> union(PCollection<T> rightCollection) {
+    checkNotNull(rightCollection, "rightCollection argument is null");
+    SerializableBiFunction<Long, Long, Long> unionFn = (numberOfElementsinLeft, numberOfElementsinRight) -> 1L;
+    return new SetImpl<>(rightCollection, unionFn);
+  }
+
+  /**
+   * Returns a new {@code SetFns.SetImpl<T>} transform that compute the unionAll with provided
+   * {@code PCollection<T>}.
+   *
+   * <p>The argument should not be modified after this is called.
+   *
+   * <p>The elements of the output {@link PCollection<T>} which will follow UNION_ALL semantics as
+   * follows: Given there are m elements on pipeline which is constructed {@link PCollection<T>}
+   * (left) and n elements on in provided {@link PCollection<T>} (right): - it will output m
+   * elements of left and m elements of right.
+   *
+   * <pre>{@code
+   * Pipeline p = ...;
+   *
+   * PCollection<String> left = p.apply(Create.of("1", "2", "3", "4", "5"));
+   * PCollection<String> right = p.apply(Create.of("1", "3", "4", "6"));
+   *
+   * PCollection<String> results =
+   *     left.apply(SetFns.unionAll(right));
+   * }</pre>
+   */
+  public static <T> SetImpl<T> unionAll(PCollection<T> rightCollection) {
+    checkNotNull(rightCollection, "rightCollection argument is null");
+    SerializableBiFunction<Long, Long, Long> unionFn = Long::sum;
+    return new SetImpl<>(rightCollection, unionFn);
+  }
+
+  private static <T> PCollection<T> performSetOperation(
+      PCollection<T> leftCollection,
+      PCollection<T> rightCollection,
+      SerializableBiFunction<Long, Long, Long> fn) {
+
+    TupleTag<Void> leftCollectionTag = new TupleTag<>();
+    TupleTag<Void> rightCollectionTag = new TupleTag<>();
+
+    MapElements<T, KV<T, Void>> elementToVoid =
+        MapElements.via(
+            new SimpleFunction<T, KV<T, Void>>() {
+              @Override
+              public KV<T, Void> apply(T element) {
+                return KV.of(element, null);
+              }
+            });
+
+    PCollection<KV<T, Void>> left = leftCollection.apply("PrepareLeftKV", elementToVoid);
+    PCollection<KV<T, Void>> right = rightCollection.apply("PrepareRightKV", elementToVoid);
+
+    PCollection<KV<T, CoGbkResult>> coGbkResults =
+        KeyedPCollectionTuple.of(leftCollectionTag, left)
+            .and(rightCollectionTag, right)
+            .apply(CoGroupByKey.create());
+
+    return coGbkResults.apply(
+        ParDo.of(
+            new DoFn<KV<T, CoGbkResult>, T>() {
+
+              @ProcessElement
+              public void processElement(ProcessContext c) {
+                KV<T, CoGbkResult> elementGroups = c.element();
+
+                CoGbkResult value = elementGroups.getValue();
```

Comment:
I wonder if it's worth dropping at TODO here to lift combiners through the CoGBK.

Reply:
Thanks. I have put a comment. can you clarify please? You mean use `CoGbk.compose`? If possible I would implement it instead of TODO if is not much of an effort.

Reply:
@robertwb : Do we need this comment? I not understand what is purpose of this TODO quiet well.